transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -163,166 +160,82 @@ def _try_connect(self, connection_string):
             except:
                 raise
 
-            self._control_thread = JLinkControlThread(self.jlink)
-            self._control_thread.start()
-
             self.set_config('probe_required', True)
             self.set_config('probe_supported', True)
 
-    def stop_sync(self):
-        """Synchronously stop this adapter and release all resources."""
+    async def stop(self):
+        """Asynchronously stop this adapter and release all resources."""
 
-        if self._control_thread is not None and self._control_thread.is_alive():
-            self._control_thread.stop()
-            self._control_thread.join()
+        await self._jlink_async.close_jlink()
 
-        if self.jlink is not None:
-            self.jlink.close()
+    async def probe(self):
```

Comment:
The `probe` API function can be called without a connection established and should not fail even if the connected device has blank flash or is hard-faulted, etc.  The usage in CoreTools is:

- For device adapters that don't gratuitously send `device_seen` events, they set a probe_required flag saying they need to polled to see if a device is present. 
- whenever a user calls `scan` on `HardwareManager` and the device adapter requires probing, it will call `probe`.  `probe` should ensure that any connected devices generate a `device_seen` event before returning (as you do below)
- if there are no connected devices that have a uuid then probe returns without error (since the probe worked) but no device_seen are generated.

So if there is no control info block found, probe shouldn't fail, it should just not generate any `device_seen` events.  Also, probe cannot assume anything about the status of a connection.  It may be called both during a connection and when there is no connection.  If there is a connection and that somehow means that it cannot probe for a control structure (which I don't believe is the case) then it can generate a `device_seen` event based on cached data for the current connection.

Reply:
Thank you for explanation. Basically probe either reports a connected device or does nothing, am I right?

Reply:
Yes, except I want to be clear on the meaning of `connected`.  As a general point `probe` reports a device that **is available for connection**, i.e. calling `connect <UUID>` with the UUID returned by probe should be expected to work (assuming you're not already connected), or does nothing.  

For the jlink that means the device is:
- physically connected to the jlink
- has functioning firmware with a control structure in RAM to process RPCs

Reply:
I see, thank you for the answer, I updated probe so that now it tries to connected to any of KNOWNDEVICES and find control structure in RAM.